### ISSUE TITLE
Thyra: build failure on MSVC due to use of Scalar type

### ISSUE THREAD

**Issue by mplooster** [trilinos/Trilinos#1501]
Found while working on #1160.

Sometime in the last month a change in the Thyra code led to [these build failures](https://testing.sandia.gov/cdash/viewBuildError.php?buildid=2999060) on Windows Visual Studio 2015. Apparently, it has issues with the Scalar type.

Thyra isn't terribly critical right now for #1160 (it was only being built as an optional package), so I will just turn it off and continue my work. I'm mostly creating the issue to let @trilinos/thyra know about it and in case it needs to be revisited in the future. Hopefully it will clear up as Thrya development continues, but if it doesn't we at least know a rough time frame of when the code that caused this build failure changed.

**Comment by mhoemmen** [comment]
@trilinos/thyra

**Comment by bartlettroscoe** [comment]
@atoth1, this is likely due to the Thyra/GPU refactoring described in #330.  Can you please take a look?

CC: @rppawlo 

P.S. I never got an email about this issue.  Something must be broken about the GitHub notification emails or something.

**Comment by atoth1** [comment]
@bartlettroscoe Yeah, I can try to look into this. I don't have a Windows machine or Visual Studio, so what's the best way for me to try to replicate these build errors/test a fix?

**Comment by atoth1** [comment]
@bartlettroscoe So it seems like all the build errors that this is getting are due any to call to applyOp without the Thyra namespace specified in VectorDefaultBase and MultiVectorDefaultBase. Both these contain calls to Thyra::applyOp with the namespace specified, and those look to be compiling fine. All the code is wrapped in the Thyra namespace, so I'm not sure why this distinction would matter. So my initial thought is that prepending the namespace to the applyOp calls could fix this, but I'm still not sure how/where to test this out.

**Comment by bartlettroscoe** [comment]
@atoth1,

> So it seems like all the build errors that this is getting are due to any call to applyOp without the Thyra namespace specified in VectorDefaultBase and MultiVectorDefaultBase. Both these contain calls to Thyra::applyOp with the namespace specified, and those look to be compiling fine. All the code is wrapped in the Thyra namespace, so I'm not sure why this distinction would matter. So my initial thought is that prepending the namespace to the applyOp calls could fix this, but I'm still not sure how/where to test this out.

I would just add `Thyra::` to the beginning of these and push with the ]checkin-test-sems.sh](https://github.com/trilinos/Trilinos/wiki/Policies-%7C-Safe-Checkin-Testing) script.  Then you can see if the problem goes away the next day.

**Comment by atoth1** [comment]
@mplooster Would you be able to try out this patch to see if it cleans up these build errors? I'd like to know whether this is actually fixes the issue before I push it in.
[0001-Thyra-Prepending-namespace-to-applyOp-calls-fix-for-.txt](https://github.com/trilinos/Trilinos/files/1160507/0001-Thyra-Prepending-namespace-to-applyOp-calls-fix-for-.txt)

**Comment by mplooster** [comment]
@atoth1 on my local machine, that seems to clean up the build errors.

**Comment by atoth1** [comment]
Great, I'll get this pushed in and monitor the dashboard to see if things get fixed there.

**Comment by atoth1** [comment]
@bartlettroscoe I just tried the checkin-test script, and I guess I don't have push access for this. Should I set up a PR for this, or can I get added to the list for push access?

**Comment by bartlettroscoe** [comment]
@atoth1, I did not see this until now (not getting many GitHub notification emails).

You can submit as a PR and I will review and push for you.

**Comment by bartlettroscoe** [comment]
I just pushed the commit 7577e04.

@mplooster, please let us know if this fixes the issue.

Putting in review..

**Comment by mplooster** [comment]
There is still a build issue, but I'm not sure if it related or not. See [here](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3036198) and [here](https://testing.sandia.gov/cdash/index.php?project=Trilinos&parentid=3036177) for the details.

**Comment by bartlettroscoe** [comment]
@mplooster, the build error for Thyra shown at:

* https://testing.sandia.gov/cdash/viewBuildError.php?buildid=3036212

shows:

```
Teuchos_StandardUnitTestMain.cpp  C:\Users\mgploos\TrilinosProject\MPI_Debug_12.11\Trilinos\packages\teuchos\core\test\UnitTest\Teuchos_StandardUnitTestMain.cpp : fatal error C1083: Cannot open compiler generated file: '': Invalid argument
--
```

What does that mean?

**Comment by mplooster** [comment]
@bartlettroscoe That error comes when it is trying to build ThyraCore_DiagonalQuadraticResponseOnlyModelEvaluatorUnitTests. Without having looked at it closely, I suspect there might be an issue with setting up the build for that test. I can't claim tons of familiarity with Teuchos_StandardUnitTestMain or Thyra, so in all honesty I really don't know. Is Thyra doing something different than other packages when it is trying to build its unittests? Also, I am wondering if something else changed in Thyra between the time I tested the patch on my local machine 18 days ago (when I didn't get this error), and now.

It is a different build error than the one this ticket was created for, so do you want a different ticket opened for it?

**Comment by bartlettroscoe** [comment]
@mplooster,

This sounds like a new failure and therefore should be a new GitHub Issue.  Without being able to reproduce this error myself, it would be extremely difficult to figure out what this compiler is doing.   What version of MSVC is this?

Assuming that I will never be able to reproduce this problem myself,  after creating the new issue, you should disable the Thyra tests for this build using:

```
-DThyra_ENABLE_TESTS=OFF
```

If passing that to the inner configure options does not work, let me know and will figure out how to fix the problem.  Again, once the all-at-once configure, build, test, and submit feature is complete, it will simplify managing these different configuration options.

BTW, what Trilinos customer is requesting this MSVC build of Trilinos?  What parts of Trilinos that that customer actually need?

**Comment by mplooster** [comment]
@bartlettroscoe Will do. It is Visual Studio 2015. I have already disabled the Thyra tests (as I mentioned in the original ticket description). For the builds with this new error, I only turned it on temporarily to see if the patch fixed this issue. For more details about the Trilinos customer and their specific needs, please contact @jwillenbring.
